src/geometry/euclidean/circle.lean

```diff
@@ -0,0 +1,176 @@
+  ... = abs (r ^ 2 - l ^ 2) * abs (∥y∥ ^ 2) : by rw ← abs_mul _ _
+  ... = abs (r ^ 2 - l ^ 2) * ∥y∥ ^ 2       : by rw abs_of_nonneg (pow_two_nonneg ∥y∥)
+  ... = abs ((r - l) * (r + l)) * ∥y∥ ^ 2  : by rw (show r ^ 2 - l ^ 2 = (r - l) * (r + l), by ring)
+  ... = abs (r - l) * abs(r + l) * ∥y∥ ^ 2  : by rw abs_mul (r - l) (r + l)
+  ... = (∥r - l∥ * ∥r + l∥) * ∥y∥ ^ 2         : by rw [← real.norm_eq_abs _, ← real.norm_eq_abs _]
+  ... = (∥r - l∥ * ∥y∥) * (∥r + l∥ * ∥y∥)      : by ring
+  ... = ∥(r - l) • y∥ * ∥(r + l) • y∥        : by rw [← norm_smul (r - l) y, ← norm_smul (r + l) y]
+  ... = ∥r • y - l • y∥ * ∥r • y + l • y∥    : by rw [sub_smul r l y, add_smul r l y]
+  ... = ∥r • y - y∥ * ∥r • y + y∥            : by rw hl_smul_y_eq_y
+  ... = ∥x - y∥ * ∥x + y∥                    : by rw ← hxy,
+end
+
+end inner_product_geometry
+
+namespace euclidean_geometry
+/-!
+### Geometrical results on circles in Euclidean affine spaces
+
+This section develops some results on circles in Euclidean affine spaces.
+-/
+
+open inner_product_geometry
+
+variables {P : Type*} [metric_space P] [normed_add_torsor V P]
+include V
+
+/-- If `P` is a point on the line `AB` and `Q` is equidistant from `A` and `B`, then
+`AP * BP = abs (BQ ^ 2 - PQ ^ 2)`. -/
+lemma mul_dist_eq_sub_dist_pow_two {a b p q : P} (hp : ∃ k : ℝ, k ≠ 1 ∧ b -ᵥ p = k • (a -ᵥ p))
+  (hq : dist a q = dist b q) : (dist a p) * (dist b p) = abs ((dist b q) ^ 2 - (dist p q) ^ 2) :=
+begin
+  let m : P := midpoint ℝ a b,
+  have h1 : a -ᵥ p = (m -ᵥ p) - (m -ᵥ a) := (vsub_sub_vsub_cancel_left a p m).symm,
+  have h2 : b -ᵥ p = (m -ᵥ p) + (m -ᵥ a),
+  { rw [midpoint_vsub_left a b, ← right_vsub_midpoint a b, add_comm, vsub_add_vsub_cancel b m p] },
+  have h3 : a -ᵥ q = (m -ᵥ q) - (m -ᵥ a) := (vsub_sub_vsub_cancel_left a q m).symm,
+  have h4 : b -ᵥ q = (m -ᵥ q) + (m -ᵥ a),
+  { rw [midpoint_vsub_left a b, ← right_vsub_midpoint a b, add_comm, vsub_add_vsub_cancel b m q] },
+  have h5 : p -ᵥ q = (m -ᵥ q) - (m -ᵥ p) := (vsub_sub_vsub_cancel_left p q m).symm,
+
+  rw [dist_eq_norm_vsub V a p, dist_eq_norm_vsub V b p, h1, h2],
+  rw [dist_eq_norm_vsub V b q, dist_eq_norm_vsub V p q, h4, h5],
+  rw [dist_eq_norm_vsub V a q, dist_eq_norm_vsub V b q, h3, h4] at hq,
+  rw [h1, h2] at hp,
+  exact mul_norm_eq_sub_norm_pow_two hp hq,
+end
+
+/-- If `A`, `B`, `C`, `D` are cospherical and `P` is on both lines `AB` and `CD`, then
+`AP * BP = CP * DP` .-/
+theorem mul_dist_eq_mul_dist_of_cospherical
+  {a b c d p : P} (h : cospherical ({a, b, c, d} : set P))
+  (hapb : ∃ k₁ : ℝ, k₁ ≠ 1 ∧ b -ᵥ p = k₁ • (a -ᵥ p))
+  (hcpd : ∃ k₂ : ℝ, k₂ ≠ 1 ∧ d -ᵥ p = k₂ • (c -ᵥ p)) :
+  (dist a p) * (dist b p) = (dist c p) * (dist d p) :=
+begin
+  set circle : set P := {a, b, c, d} with circle_def,
+  obtain ⟨q, r, h⟩ := (cospherical_def circle).mp h,
+  have ha₀ : a ∈ circle, finish,
+  have hb₀ : b ∈ circle, finish,
+  have hc₀ : c ∈ circle, finish,
+  have hd₀ : d ∈ circle, finish,
```

Comment:
The `finish`es seem to make this proof extremely slow. May I suggest:
```
begin
  obtain ⟨q, r, h'⟩ := (cospherical_def {a, b, c, d}).mp h,
  obtain ⟨ha, hb, hc, hd⟩ := ⟨h' a _, h' b _, h' c _, h' d _⟩,
  { rw ← hd at hc,
    rw ← hb at ha hd,
    rw [mul_dist_eq_sub_dist_pow_two hapb ha, mul_dist_eq_sub_dist_pow_two hcpd hc, hd] },
  all_goals { simp },
end
```

(You had two `h`s so I replaced the latter with `h'`)